File: src/main/java/com/bruno/matchers/MissingLettersMatcher.java
package com.bruno.matchers;

import java.util.function.Predicate;
import java.util.regex.Pattern;

public class MissingLettersMatcher implements Predicate<String> {

    private final Predicate<String> sameLengthMatcher;
    private final Pattern wordPattern;


    public MissingLettersMatcher(String word) {
        this.sameLengthMatcher = MatcherFactory.getInstance().withSameSize(word);

        String replaced = word.replaceAll("\\_", "\\.");
        this.wordPattern = Pattern.compile(replaced, Pattern.CASE_INSENSITIVE);
    }

    public boolean test(String word) {
        return this.sameLengthMatcher.test(word) && this.wordPattern.matcher(word).find();
    }
}

File: src/main/java/com/bruno/matchers/MissingLettersMatcher.kt
package com.bruno.matchers

import java.util.function.Predicate
import java.util.regex.Pattern

class MissingLettersMatcher(word: String) : SameLengthMatcher(word), Predicate<String> {

    private val wordPattern: Pattern = word.let {
        it.replace("\\_".toRegex(), "\\.")
    }.let {
        Pattern.compile(it, Pattern.CASE_INSENSITIVE)
    }

    override fun test(word: String): Boolean {
        return super.test(word) && this.wordPattern.matcher(word).find()
    }
}
